crates/contracts/solidity/Trader.sol

```diff
@@ -21,8 +21,8 @@ contract Trader {
     address constant private TRADER_IMPL = address(0x10000);
 
     /// @dev Flag that ensures that `prepareSwap` gets called exactly once to
-    /// prevent custom interaction from calling it.
-    bool private _alreadyCalled = false;
+    /// prevent custom interactions from calling it.
+    address private _prepareSwapMarker;
```

Comment:
A more common implementation pattern to deal with this kind of thing is as follows:

```solidity
bytes32 constant private ALREADY_CALLED_SLOT = keccak256("Trader.alreadyCalled") - 1;
```

And then:

```solidity
function alreadyCalled() private returns (bool value) {
    bytes32 slot = ALREADY_CALLED_SLOT;
    assembly {
        value := sload(slot)
        sstore(slot, 1)
    }
}
```

And finally:

```solidity
require(!alreadyCalled(), "prepareSwap can only be called once");
```

Reply:
The reason for this suggestion is that it *may* be that `ReentrencyGuard` relies on specific values in that storage slot.

Reply:
Thanks! Will try it out. What purpose does `-1` serve when computing the slot here?

Reply:
[52b3ff7](https://github.com/cowprotocol/services/pull/1735/commits/52b3ff78b24b6b81c7b4d695c74abefa137a42b6)